internal/backend/remote-state/s3/client.go

```diff
@@ -334,27 +387,94 @@ func (c *RemoteClient) Unlock(id string) error {
 	ctx, baselog := baselogging.NewHcLogger(ctx, log)
 	ctx = baselogging.RegisterLogger(ctx, baselog)
 
-	if c.ddbTable == "" {
+	if !c.useLockFile && c.ddbTable == "" {
 		return nil
 	}
 
 	lockErr := &statemgr.LockError{}
 
 	log.Info("Unlocking remote state")
 
+	if c.useLockFile {
+		err := c.unlockWithFile(ctx, id, log)
+		if err != nil {
+			lockErr.Err = err
+			return lockErr
+		}
+
+		return nil
+	} else {
+		err := c.unlockWithDynamoDB(ctx, id, lockErr, log)
+		if err != nil {
+			lockErr.Err = err
+			return lockErr
+		}
+	}
+
+	return nil
+}
+
+// unlockWithFile attempts to unlock the remote state by deleting the lock file from Amazon S3.
+//
+// This method is used when the S3 native locking mechanism is in use, which uses a `.tflock` file
+// to manage state locking. The function deletes the lock file to release the lock, allowing other
+// Terraform clients to acquire the lock on the same state file.
+func (c *RemoteClient) unlockWithFile(ctx context.Context, id string, log hclog.Logger) error {
+	// Retrieve the lock file to get its ETag, which is used to verify the lock ID.
+	getOutput, err := c.s3Client.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(c.bucketName),
+		Key:    aws.String(c.lockFilePath),
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to retrieve lock file: %w", err)
+	}
+
+	// Clean up extra quotes from the ETag value.
+	// ETag values from S3 are often quoted, so we remove the quotes to ensure accurate comparison.
```

Comment:
Note for self, they are not 'often' quoted, in the GetObjectResponse the ETag is always double quoted.

Reply:
fixed in [eddf8f8](https://github.com/hashicorp/terraform/pull/35661/commits/eddf8f8b4cd5214a9b66e065162cafdcbba04751)